tests/unittests/Foundation/NSUndoManagerTests.mm

```diff
@@ -0,0 +1,718 @@
+@property (retain) NSMutableArray* arrayOfNumbers;
+- (id)initWithValue:(NSNumber*)num;
+- (id)initWithEmptyArray;
+- (void)updateNumber:(NSNumber*)value;
+- (void)updateNumberSingleUndoStep:(NSNumber*)value;
+- (void)addToMutableArray:(NSNumber*)value;
+- (void)removeFromMutableArray:(NSNumber*)value;
+@end
+
+@implementation UndoTestObject
+
+- (id)initWithValue:(NSNumber*)num {
+    self = [super init];
+    _undoManager = [[NSUndoManager alloc] init];
+    _value = [num retain];
+    _arrayOfNumbers = nil;
+    return self;
+}
+
+- (id)initWithEmptyArray {
+    self = [super init];
+    _undoManager = [[NSUndoManager alloc] init];
+    _value = nil;
+    _arrayOfNumbers = [[NSMutableArray array] retain];
+    return self;
+}
+
+- (void)dealloc {
+    [_undoManager release];
+    [_value release];
+    [_arrayOfNumbers release];
+}
+
+- (void)updateNumber:(NSNumber*)newValue {
+    [self.undoManager registerUndoWithTarget:self selector:@selector(updateNumber:) object:self.value];
+    self.value = newValue;
+}
+
+- (void)updateNumberSingleUndoStep:(NSNumber*)value {
+    [self.undoManager beginUndoGrouping];
+    [self updateNumber:value];
+    [self.undoManager endUndoGrouping];
+}
+
+- (void)addToMutableArray:(NSNumber*)value {
+    [self.undoManager registerUndoWithTarget:self selector:@selector(removeFromMutableArray:) object:value];
+    [self.arrayOfNumbers addObject:value];
+}
+
+- (void)removeFromMutableArray:(NSNumber*)value {
+    [self.undoManager registerUndoWithTarget:self selector:@selector(addToMutableArray:) object:value];
+    [self.arrayOfNumbers removeObject:value];
+}
+
+@end
+
+TEST(NSUndoManager, BasicUndoTest) {
+    NSNumber* oldValue = [NSNumber numberWithInt:5];
+    NSNumber* newValue = [NSNumber numberWithInt:6];
+
+    UndoTestObject* object = [[[UndoTestObject alloc] initWithValue:oldValue] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
+    [object updateNumber:newValue];
+
+    EXPECT_EQ([object.undoManager canUndo], YES);
+    EXPECT_EQ(object.value.intValue, newValue.intValue);
+
+    [object.undoManager undo];
+
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
+    EXPECT_EQ([object.undoManager canUndo], NO);
+}
+
+TEST(NSUndoManager, MultipleUndoTests) {
+    NSNumber* val1 = [NSNumber numberWithInt:5];
+    NSNumber* val2 = [NSNumber numberWithInt:6];
+    NSNumber* val3 = [NSNumber numberWithInt:7];
+    NSNumber* val4 = [NSNumber numberWithInt:8];
+
+    UndoTestObject* object = [[[UndoTestObject alloc] initWithValue:val1] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    // Each undo operation here gets added to the single default undo group.
+    [object updateNumber:val2];
+    [object updateNumber:val3];
+    [object updateNumber:val4];
+    EXPECT_EQ(object.value.intValue, val4.intValue);
+    [object.undoManager undo];
+    EXPECT_EQ(object.value.intValue, val1.intValue);
```

Comment:
Because groupsByEvent is on by default, a top level undo group is opened and closed automatically.

---
In reply to: [77102647](https://github.com/Microsoft/WinObjC/pull/900#discussion_r77102647) [](ancestors = 77102647,77097544)